Latest import csv

timifasubaa opened the pull request:
This PR allows users to import CSV files to superset. It extends the work that had already been done  in (https://github.com/apache/incubator-superset/pull/2381) and also adds the new functionality of uploading to a hive datasource. All feedback to make the code and user experience better is welcome.

@mistercrunch

Commit:
Testing gitlab remote repo

Ryan4815 committed:
test commit

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Commit:
Merge remote-tracking branch 'origin/master' into csv-import

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Commit:
trying out various ways of uploading xls/csv with flask

Commit:
Merge remote-tracking branch 'origin/master' into csv-import

Ryan4815 committed:
revert previous changes

Ryan4815 committed:
Add CSV button placed on database page

Ryan4815 committed:
Changed upload csv to use same layout/template as import dashboards

Ryan4815 committed:
Starting to add csv to db form (to expose panda api)

Ryan4815 committed:
CsvToDatabaseView updated to include most of the pandas api read_csv fields.

Ryan4815 committed:
Exposed the pandas .to_sql API in the upload csv form. Removed the separate upload page.

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Ryan4815 committed:
File select part of csvtodatabaseview form

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Ryan4815 committed:
If cause to set the header column to row 0 if name is none

Ryan4815 committed:
Merge remote-tracking branch 'origin/master' into csv-import

Ryan4815 committed:
Used flask-babel lazy_gettext to ready the new strings for translation.

Ryan4815 committed:
Moved Csv form from views.py to forms.py and removed unused imports

Ryan4815 committed:
Changed BooleanField to use BetterBooleanField

Ryan4815 committed:
Added FAQ entry on uploading csv

Ryan4815 committed:
Merge branch 'master' of https://github.com/airbnb/superset

Ryan4815 committed:
Merge remote-tracking branch 'origin/master' into csv-import

Commit:
Merge Ryan4815's fork into my fork

Commit:
Fix merge conflict

Commit:
Add tests to CSV-import

Commit:
Change the default of replace to append

Commit:
Add additional form validations

Commit:
Allow for large csv files exceeding machine memory

Commit:
Fix filepath bug

Commit:
Fix linting errors

Commit:
fix linting errors

Commit:
Remove extraneous csv file after use

Commit:
Remove extraneous file at the end of test

Commit:
Fix more linting errors

Commit:
Change port back to default

Commit:
Fix linting errors

Commit:
Fix linting error

Commit:
Add babel files to gitignore

Commit:
remove babel files from gitignore

Commit:
Add localizations and finish merge

Commit:
Fix issue remaining from previous merge

Commit:
Fix previous merge issues.

Commit:
Add missing file.

Commit:
Merge branch 'master' of https://github.com/airbnb/superset into csv-import-dev

Commit:
Add unicodecsv to setup.py

Commit:
test commit

Commit:
Fix codeclimate and travis issues

Commit:
Merge branch 'master' of https://github.com/airbnb/superset into csv-import

Commit:
Merge branch 'csv-import-travis' into csv-import

SalehHindi committed:
Fix merge conflicts in translation files

SalehHindi committed:
Merge remote-tracking branch 'upstream/master' into csv-import

SalehHindi committed:
Add newly imported csv table to the list of table models.

SalehHindi committed:
Sync fork with upstream

mistercrunch committed:
merged

mistercrunch committed:
CSV LATEST

mistercrunch committed:
db_engine

timifasubaa committed:
db_engine_specs first stab

migrate the upload logic to dbengine_specs

xrmx committed:
Miscvizcleanups (#3476)

* viz: remove unused countries import

* viz: simplify code in BaseViz.query_obj

xrmx committed:
docs: athena can be installed from pypi (#3477)

As sqlalchemy support has been merged and available in a release
since a few.

xrmx committed:
viz: fix reversed stats_logger label (#3475)

mistercrunch committed:
Better installation docs (#3469)

JulieRossi committed:
Create CsvResponse to manage csv exports encoding (#3484)

dmigo committed:
Feature: "Impersonate user" setting on Datasource (#3404)

* Add "Impersonate user" setting to Datasource

* Add tests

* Use g.user.username for all the sync cases

* use uri.username instead of uri.user

* Small refactoring

datinho committed:
Getting datatype with its dialect (#3486)

fabianmenges committed:
Feature/Fix: Get a full times_series for your filter instead of Topn for each point in time (#3434)

small fix

timifasubaa committed:
Improve code quality (#3480)

* add lanscape?

* add code climate badges

* pylint first pass

* Try out yapf

* merge

* merge

* lint

* more yapf

* removed unnecessary corrections

mistercrunch committed:
[heatmap] account for bounds = 0 (#3474)

* [heatmap] account for bounds = 0

* Fix sorting

* linting

mistercrunch committed:
Removing dependency on pythrifthiveapi (#3494)

Since the latest pyhive, we don't need pythrifthiveapi as they ship with
the latest version.

There's actually a conflict between the new pyhive and pythrifthiveapi
and this fixes it.

mistercrunch committed:
Adding missing future imports (#3493)

mistercrunch committed:
Removing super() call from refactor (#3500)

timifasubaa committed:
move implementation to db_engine_spec

nit

...

..

trying static

...

timifasubaa committed:
update contributing.md (#3495)

Uses pip instead to setup superset.

timifasubaa committed:
Fix idna requirement (#3497)

There is often conflicting versions of idna. This fixes is to 2.5.

timifasubaa committed:
Remove repeated line (#3491)

* add lanscape?

* Revert "add lanscape?"

This reverts commit b3d6e80af49120b030f7d3d5eff0592a1e0566f2.

* remove unused line

Commit:
js translation -- performance improvment (#3390)

* Chinese page

* Using react-intl-universal to improve multi language in react page

* Using react-intl-universal to improve multi language in react page

* react_intl_universal

* change

* change

* change

* change

* change

* change

* change

* merge

* multiple page in js

* merge

* merge

* merge

* merge

* Js Translations

* JS Translation

* JS Translations

* Js translation

* JS translations

* JS translations

* Js translaion

* JS en Translation

* JS Translation

* upgrade document

Fixing the damn build (#3179)

* Fixing the build

* Going deeper

[bugfix] only filterable columns should show up in FilterBox list (#3105)

* [bugfix] only filterable columns should show up in FilterBox list

* Touchups

Datasource cannot be empty (#3035)

add title description to model view (#3045)

* add title description to model view

* add missing import

Add 'show/hide totals' option to pivot table vis (#3101)

[bugfix] numeric value for date fields in table viz (#3036)

Bug was present only when using the NOT GROUPED BY option

fixes https://github.com/ApacheInfra/superset/issues/3027

fix hive.fetch_logs (#2968)

add Zalando to the list of organizations (#3171)

docs: fixup installation examples code indentation (#3169)

[bugfix] fix bar order (#3180)

[bugfix] visualize flow error: 'Metric x is not valid' (#3181)

The metric name in the frontend doesn't match the one generated on the
backend. It turns out the explore view will default to the first
metric so specifying one isn't needed.

Fix the segment interval for pulling metadata (#3174)

The end of the interval would be on the truncated today date, which
means that you will exclude today. If your realtime ingestion job
runs shorter than a day, the metadata cannot be pulled from the
druid cluster.

Bump cryptography to 1.9 (#3065)

As 1.7.2 doesn't compile here with openssl 1.1.0f

Escaping the user's SQL in the explore view (#3186)

* Escaping the user's SQL in the explore view

When executing SQL from SQL Lab, we use a lower level API to the
database which doesn't require escaping the SQL. When going through
the explore view, the stack chain leading to the same method may need
escaping depending on how the DBAPI driver is written, and that is the
case for Presto (and perhaps other drivers).

* Using regex to avoid doubling doubles

[sqllab] improve Hive support (#3187)

* [sqllab] improve Hive support

* Fix "Transport not open" bug
* Getting progress bar to show
* Bump pyhive to 0.4.0
* Getting [Track Job] button to show

* Fix testzz

Add BigQuery engine specifications (#3193)

As contributed by @mxmzdlv on issue #945

[bugfix] fix merge conflict that broke Hive support (#3196)

Adding 'apache' to docs (#3194)

[druid] Allow custom druid postaggregators (#3146)

* [druid] Allow custom druid postaggregators

Also, fix the postaggregation for approxHistogram quantiles so it adds
the dependent field and that can show up in the graphs/tables.

In general, postAggregators add significant power, we should probably
support including custom postAggregators. Plywood has standard
postAggregators here, and a customAggregator escape hatch that allows
you to define custom postAggregators.

This commit adds a similar capability for Superset and a additional
field/fields/fieldName breakdown of the typical naming for dependent
aggregations, which should make it significantly easier to develop
approxHistogram and custom postAggregation-required dashboards.

* [druid] Minor style cleanup in tests file.

* [druid] Apply code review suggestions

* break out CustomPostAggregator into separate class. This just cleans
  up the creation of the postaggregator a little bit.
* minor style issues.
* move the function around so the git diff is more readable

add combine config for metrics in pivot table (#3086)

* add combine config for metrics in pivot table

* change method to stack/unstack

* update backendSync

Autofocus search input in VizTypeControl modal onEnter (#2929)

Speed up JS build time (#3203)

Also bumping a few related libs

JS Translation

JS translations

js translation

fix issue 3204 (#3205)

[bugfix] capture Hive job_id pre-url transformation (#3213)

js translation

fix issue 3204 (#3205)

[bugfix] capture Hive job_id pre-url transformation (#3213)

[docs] update url in CONTRIBUTING.md (#3212)

[sqllab/cosmetics] add margin-top for labels in query history (#3222)

[explore] nvd3 sort values in rich tooltip (#3197)

[sqllab] fix UI shows 'The query returned no results' momentarily (#3214)

this is visible when running async queries between the fetching and
success state as the rows are getting cached in the component

[explore] DatasourceControl to pick datasource in modal (#3210)

* [explore] DatasourceControl to pick datasource in modal

Makes it easier to change datasource, also makes it such that the list
of all datasources doesn't need to be loaded upfront.

* Adding more metadata

* Js translation

* js tran

* js trans

* js trans

* js tran

* js trans

* js trans

* js tran

* js translation

* js trans

* js translation

* try load language pack async

* Backend translations things

* create language pack inside common data

* performance improvement for js i18n.

- js bundle should not contain localized content
- we populate translation content from server-side, in boostrap.common.language_pack
- in client-side, use promise to wrap around translation content. text will be translated after translation content arrived/parsed.
- fix linting

* fix Timer unit test

* 1. add global hook for all tests, to make translation pack avaialble before each test starts.
2. fix unit test for Timer component
3. remove noused method get_locale, and modules
4. fix page reload after user change page language

* parse and build i18n dictionary as a module

* fix sync-backend task, which should run without DOM

michellethomas committed:
Add metric warning (#3499)

* Adding warning text to metrics

* Adding javascript tests

* Fixing downgrade script for warning_text

* Adding merge migration

Mogball committed:
Add Table performance improvements (#3509)

* Improved performance of 'Add table' function

* got rid of pvt function call

* changes metric obj to key on metric_name

Commit:
allow user update slice name in dashboard view (#3467)

- if current user is allowed to edit dashboard, we will allow this user to edit slice name.
- show different tooltip given allowed/not-allowed to update slice name.
- user will click slice name and update.
- after user submit edit, if he doesn't have right to alter slice, server-side will return error message to client-side. Slice name will not be changed or saved.
- will show notification after save slice name.

Commit:
Allow user update slice title in visualize flow (#3466)

* 1. after user make sql query and visualize, allow user click title to update slice title, and create a new slice at the same time.
2. don't save new title if it is empty. Will still show old title.

* change saveSlice call response and update explore view

timifasubaa committed:
working with dropdown and correct writes

timifasubaa committed:
update faq

...

...

add the config options

make columns show

update

...

Mogball committed:
Druid refresh metadata performance improvements (#3527)

* parallelized refresh druid metadata

* fixed code style errors

* fixed code for python3

* added option to only scan for new druid datasources

* Increased code coverage

fabianmenges committed:
Feature: Display the verbose name for metrics within Charts and legend. (#3504)

Mogball committed:
Feature: query string API endpoint (#3513)

* exposed API endpoint to get querystring for a slice

* Added unit tests for endpoint

* fixed test case for python3

* moved get querystring logic into its own func

* renamed query string endpoint

Mogball committed:
Feature: Paired t-test table visualization (#3473)

* Feature: paired t-test table viz

* Added unit tests for viz

michellethomas committed:
Break word on InfoTooltip (#3532)

* Break word on InfoTooltip

* Moving style object to separate const

mistercrunch committed:
[style] no bold on dashboard widget headers (#3531)

xrmx commented:
I think the approach of a pandas datasource as in #3492 is more promising

xrmx committed:
docs: QUERY_TIMEOUT_THRESHOLD is gone (#3537)

mistercrunch commented on superset/__init__.py:
```diff
@@ -156,6 +156,7 @@ def index(self):
 
 get_session = appbuilder.get_session
 results_backend = app.config.get("RESULTS_BACKEND")
+csv_upload_backend = app.config.get("CSV_UPLOAD_BACKEND")
```
unused?

mistercrunch commented on superset/config.py:
```diff
@@ -289,11 +293,21 @@ class CeleryConfig(object):
+# if enabled, it can be used to store external Hive tables on the
+# S3 location specified in CSV_UPLOAD_BUCKET
+#from S3Cache.s3cache import S3Cache
+CSV_UPLOAD_BACKEND = None #if not CSV_TO_HIVE_UPLOAD_BUCKET else S3Cache(CSV_TO_HIVE_UPLOAD_BUCKET, "EXTERNAL_HIVE_TABLES")
```
ununsed?

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+                                 iterator=True)
+        df = pandas.DataFrame()
+        df = pandas.concat(chunk for chunk in chunks)
+        print(df)
```
nope

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+
+    @staticmethod
+    def upload_csv(form):
+        print(results_backend)
```
let's not leave all these print statements around. Use `logging.info` if you want to log things

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+        print(results_backend)
+        def allowed_file(filename):
+            # Only allow specific file extensions as specified in the config
+            return '.' in filename and \
```
using index `[-1]` would allow for filenames with multiples dots in them, but better is to use the `os.path` module as in https://docs.python.org/2/library/os.path.html#os.path.splitext

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +81,103 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+                            escapechar=form.escapechar.data,
+                            comment=form.comment.data,
+                            error_bad_lines=form.error_bad_lines.data,
+                            chunksize=10000)
```
seems arbitrary, unclear where it belongs at this point though, maybe a database configuration element? Ok leaving as is for now though.

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+        #TODO(timifasubaa): replace the current approach with the results backend approach above.
+
+        bucket_path = config["HIVE_CSV_BUCKET_PATH"] 
+        s3_dir = "hive_csv/"+table_name+"/"
```
PEP8, lint your diff, also use `os.path.join`

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+        bucket_path = config["HIVE_CSV_BUCKET_PATH"] 
+        s3_dir = "hive_csv/"+table_name+"/"
+
+        upload_path = config['UPLOAD_FOLDER'] + secure_filename(form.csv_file.data.filename)
```
`os.path.join`

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+        import boto3
+        s3 = boto3.client('s3')
+        s3.upload_file(upload_path, 'airbnb-superset', s3_dir+file_name)
+        sql = "CREATE EXTERNAL TABLE " + str(table_name) + " ( " + str(schema_definition) + " ) " +\
```
use a triple quoted string with `.format()` for long string assembly as in:
```
s = """
    CREATE EXTERNAL TABLE {table}
   ...
""".format(**locals())
```

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +790,41 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+            engine = create_engine(form.con.data)
+            engine.execute(sql)
+        except Exception as e:
+            print(e)
```
Also, that means all errors get silenced. Errors need to bubble up to users.

mistercrunch commented on superset/forms.py:
```diff
@@ -0,0 +1,1367 @@
+"""Contains the logic to create cohesive forms on the explore view"""
```
Does this huge file belong in your PR? Seems like an artifact from explore v1.

mistercrunch commented on superset/translations/it/LC_MESSAGES/messages.po:
```diff
@@ -2789,8 +2789,8 @@ msgstr ""
 msgid ""
-"Datetime column not provided as part table configuration and is required "
-"by this type of chart"
+"Defines the origin where time buckets start, accepts natural dates as in "
```
Let's keep all of the translations out of this PR

mistercrunch commented on superset/views/core.py:
```diff
@@ -12,13 +12,14 @@
 import traceback
-
+import os
+import pandas
```
unused?

mistercrunch replied:
btw pylint should tell you when you have unused imports

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
+        form.index_label.data = None
+
+    def form_post(self, form):
+        # post process form
```
method name is self-explanatory, comment is confusing or doesn't add value

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
+    def form_post(self, form):
+        # post process form
+
+        # Turn into list of strings
```
unrelated?

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
+
+        if form.names.data is not None:
+            form.names.data = form.names.data.split(",")
+        else:
```
elif?

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
+        filename = upload_file(form.csv_file.data)
+
+        # some of this stuff belongs in db_engine_spec
+        #print(form.data)
```
please clean up those artifacts, makes me feel like this PR isn't ready for review

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
+
+        # some of this stuff belongs in db_engine_spec
+        #print(form.data)
+        database = db.session.query(models.Database).filter_by(sqlalchemy_uri=form.data.get('con')).one()
```
line too long, try to respect the <=80-90 PEP8 guideline, you can use js-like method chaining syntax between parenthesis in python as in:
```
database = ( 
    db.session.query(models.Database)
    .filter_by(sqlalchemy_uri=form.data.get('con'))
    .one()
)
```

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,129 @@ class DatabaseAsync(DatabaseView):
+    #                           filename)
+
+class CsvToDatabaseView(SimpleFormView):
+    #DatasourceModelView,  add to ^ ?
```
clean up comment, add doctring https://www.python.org/dev/peps/pep-0257/

mistercrunch commented on tests/testCSV.csv:
```diff
@@ -0,0 +1,2 @@
+Name,TestField
```
? cleanup

mistercrunch reviewed (changes requested):
PR doesn't seem ready for review, please add a [WiP] work in progress tag when that's the case.

Also it'd be great to reference the original PR in the PR description. Github will notify people tracking that PR as a result.

timifasubaa commented:
Sorry. I should have made it clearer that it was WIP. I've now updated it. I wanted to get general feedback on the approach I took before cleaning up the code and you've done that. Thank you.

timifasubaa committed:
use S3Cache

timifasubaa committed:
fix s3cache errors

some more cleanup

timifasubaa committed:
error messaging

timifasubaa committed:
load file contents correctly

...

..

xrmx committed:
Add support for column specific fillna to viz (#3066)

Fix #3029

xrmx committed:
setup: bump pandas to 0.20.3 (#3506)

mistercrunch committed:
v0.20.0 (#3544)

mistercrunch committed:
[explore] fix cached tooltip (#3526)

Commit:
Time Series Annotation Layers (#3521)

* Adding annotations to backend

* Auto fetching Annotations on the backend

* Closing the loop

* Adding missing files

* annotation layers UI

for https://github.com/apache/incubator-superset/issues/3502

* a few fixes per code review.

- add annotation input sanity check before add and before update.
- make SelectAsyncControl component statelesis, and generic
- add annotation description in d3 tool tip
- use less variable to replace hard-coded color

Commit:
try to fix problem that chrome window not opening after ajax requrest (#3528)

timifasubaa committed:
no po file changes

...

timifasubaa committed:
remove csv_backend

timifasubaa committed:
delete form.py and move code elsewhere

mend

mend

mend

...

mend

mend

mend

mend

mend

mend

..

..

--amend

verify uploaded file is csv

timifasubaa committed:
Merge branch 'master' into latest_import_csv

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
         return {}
 
+    @staticmethod
+    def csv_to_df(names, filepath_or_buffer, sep, header, index_col, squeeze,
```
Some linters will trigger on "too many params" on methods. This has lots of boilerplate code here. We can chat about how to improve this using a data structure and an unpack operator `**`, but it may be ok as is. For instance you could have a `pd_read_csv_kwargs` dict that you'd pass around, and ultimately call `pandas.read_csv(**pd_read_csv_kwargs)`

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+
+        database = (
+            db.session
+                .query(type(query)) #Database
```
NIT: I think in most other places in the code base we line up the js-style method chaining with the first line. It's not PEP8 or anything, just within-codebase consistency.

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+            flash("Invalid file trype selected.", 'alert')
+            return False
+        df = BaseEngineSpec.csv_to_df(names=form.names.data,
+                            filepath_or_buffer=filename,
```
NIT: doesn't the linter trigger on this? `names` should be aligned with `filepath_or_buffer`, personally I'd break after the `(` and go with just 4 indents.

This is also a lot of boilerplate... There's a way to do this where the list of args is defined only once, in an array. We can chat about it.

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,115 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+                            error_bad_lines=form.error_bad_lines.data,
+                            chunksize=10000)
+
+        BaseEngineSpec.df_to_db(df=df,
```
NIT: break after `(` is your linter not triggering on this?

timifasubaa replied:
Fixed

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
 
+    @staticmethod
+    def upload_csv(form, table):
+        "uploads a cv file and creates a superset table in the right databse"
```
many (4?) typos and wrong capitalization

timifasubaa replied:
Fixed!

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+        "uploads a cv file and creates a superset table in the right databse"
+        "returns true if it was siccessful. False otherwise"
+        def get_column_names(filepath):
+            import csv
```
imports go at the top of the module, linter should trigger on this

timifasubaa replied:
Done

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+
+        
+
+        import boto3
```
import on top of module

timifasubaa replied:
Done

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+
+        import boto3
+        s3 = boto3.client('s3')
+        s3.upload_file(upload_path, 'airbnb-superset', os.path.join(upload_prefix, table_name, filename))
```
line too long

timifasubaa replied:
Done

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +805,50 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+        import boto3
+        s3 = boto3.client('s3')
+        s3.upload_file(upload_path, 'airbnb-superset', os.path.join(upload_prefix, table_name, filename))
+        sql = "CREATE EXTERNAL TABLE {} ( {} ) ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' STORED AS TEXTFILE LOCATION '{}'"\
```
as from previous comment, use triple quoted strings and kwargs as opposed to args in `.format`. Checkout the magic convention `.format(**locals())` to avoid boilerplate

timifasubaa replied:
Done

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,94 @@ class DatabaseAsync(DatabaseView):
+    def form_get(self, form):
+        # pre process form
+        # default values
+        form.name.data = None
```
Also lots of boilerplate here.

timifasubaa replied:
These set the default values. I removed the ones that weren't changing anything.

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,94 @@ class DatabaseAsync(DatabaseView):
+            if csv_file and csv_file.filename:
+                filename = secure_filename(csv_file.filename)
+                csv_file.save(os.path.join(config['UPLOAD_FOLDER'],
+                                                 filename))
```
indent?

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +304,94 @@ class DatabaseAsync(DatabaseView):
+
+        table = SqlaTable(table_name=form.name.data)
+
+        database = db.session.query(models.Database).filter_by(sqlalchemy_uri=form.data.get('con')).one()
```
line too long

mistercrunch reviewed (changes requested):
Gave up before getting to the end as the PR is still full of lint and some previous comments have not been addressed. Sharpen your linter!

timifasubaa committed:
cleanup

timifasubaa commented on superset/csvform.py:
```diff
@@ -0,0 +1,205 @@
+                     choices=all_datasources
+                       )
+    if_exists = SelectField(_('Table Exists'),
+                            description=_('If table exists do one '
```
Fixed

timifasubaa commented on superset/csvform.py:
```diff
@@ -0,0 +1,205 @@
+    try:
+      all_datasources = db.session.query(models.Database.sqlalchemy_uri, models.Database.database_name).all()
+    except Exception as e:
+      logging.info("Exception retreiving dbs: {}".format(e))
```
Because whenever I run it, it works but it breaks when the CI is running. The error it produces is: 
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: dbs [SQL: 'SELECT dbs.sqlalchemy_uri AS dbs_sqlalchemy_uri, dbs.database_name AS dbs_database_name \nFROM dbs']

It needs the dbs table that is not existent on the travis instance but the table will exist whenever someone runs superset.

mistercrunch commented on superset/forms.py:
```diff
@@ -0,0 +1,191 @@
+            FileRequired(), FileAllowed(['csv'], _('CSV Files Only!'))])
+    all_datasources = []
+    try:
+        all_datasources = db.session\
```
This should be in view scope, not in module scope. As is, it would be interpreted when the web server (re)starts

timifasubaa replied:
Fixed

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,81 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+            'if_exists': form.if_exists.data,
+            'index': form.index.data,
+            'index_label': form.index_label.data,
+            'chunksize': 10000
```
NIT: we like trailing commas, we should see if pylint can require it

mistercrunch replied:
(maybe pylint does, and flake8 doesn't)

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +771,58 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
 
+    @staticmethod
+    def upload_csv(form, table):
+        "uploads a csv file and creates a superset table in the right databse"
```
https://www.python.org/dev/peps/pep-0257/

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +771,58 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+            form.csv_file.data.filename)
+        column_names = get_column_names(upload_path)
+        schema_definition =\
+            ", ".join([col_name + " STRING " for col_name in column_names])
```
NIT: s/col_name/s/g

timifasubaa replied:
Not sure what you mean here.

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +771,58 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+            upload_path,
+            'airbnb-superset',
+            os.path.join(upload_prefix, table_name, filename))
+        sql = "CREATE EXTERNAL TABLE {table_name} ( {schema_definition} ) "\
```
NIT: 3rd comment about triple-quote strings

mistercrunch commented on superset/translations/es/LC_MESSAGES/messages.po:
```diff
@@ -7,7 +7,7 @@ msgid ""
 msgstr ""
```
damn dude. these `.po` files are still in your PR

mistercrunch commented on superset/views/core.py:
```diff
@@ -7,18 +7,18 @@
 import time
 import traceback
-
+import os
```
NIT: when there are many imports in a module, standard lib import should be together in a group (empty line between import groups), and alpha-sorted within that group preferably 

organizing imports becomes more important as there are more imports in the module

timifasubaa replied:
Fixed

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
+            flash(message, 'info')
+            return redirect('/tablemodelview/list/')
+
+
```
NIT: one empty line too many

mistercrunch commented on superset/views/core.py:
```diff
@@ -7,18 +7,18 @@
     g, request, redirect, flash, Response, render_template, Markup,
-    url_for)
-from flask_appbuilder import expose
+    url_for, send_from_directory)
```
used?

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
+        form.if_exists.data = 'append'
+        all_datasources = []
+        all_datasources = db.session.query(
+            models.Database.sqlalchemy_uri, models.Database.database_name)\
```
NIT: line break would be much better after the coma

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
+                csv_file.save(os.path.join(config['UPLOAD_FOLDER'], filename))
+                return filename
+
+        form.names.data = form.names.data.split(",") if form.names.data else 0
```
Seems like `form.names.data` should receive consistent types, not `list` or `Int` depending... Maybe `[]` or `None`?

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +301,64 @@ class DatabaseAsync(DatabaseView):
+
+        form.names.data = form.names.data.split(",") if form.names.data else 0
+
+        database = db.session.query(models.Database)\
```
Made a comment on this PR about how we like the method chaining to be wrap across lines before.
database = (
    db.session.query(models.Database)
    .filter_by(sqlalchemy_uri=form.data.get('con'))
    .one()
)

mistercrunch commented on superset/forms.py:
```diff
@@ -0,0 +1,183 @@
+
+from superset import app, db
+import superset.models.core as models
+
```
NIT: uneeded blank line

timifasubaa committed:
Merge remote-tracking branch 'upstream/master' into latest_import_csv

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,82 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+        if not allowed_file(filename):
+            flash("Invalid file type selected.", 'alert')
+            return False
+        kwargs = {
```
I don't think I was super clear about the kwargs comment earlier, and the verbose boilerplate isn't that bad either (others may consider the verbose, metaprogramming-free option better). 

Knowing the form object has a data dict:
http://wtforms.simplecodes.com/docs/0.6.1/forms.html#wtforms.form.Form.data

You can do something like:
```
form_data = form.data
csv_to_df_fields = ['names', 'filepath_or_buffer', ...]
kwargs = {s: form_data.get(s) for s in csv_to_df_fields}
kwargs['chunksize'] = 10000
# other alterations as needed
df = BaseEngineSpec.csv_to_df(**kwargs)
```

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,82 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+        return df
+
+    @staticmethod
+    def df_to_db(**kwargs):
```
if you change the function signature to `def df_to_db(df, table, con, schema, **kwargs):` then you don't need the 5 following lines and avoid juggling with kwargs

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -74,6 +84,82 @@ def extra_table_metadata(cls, database, table_name, schema_name):
+        del kwargs['df']
+        del kwargs['table']
+        kwargs['con'] = create_engine(kwargs['con'], echo=False)
+        df.to_sql(**kwargs)
```
having taken stuff out of kwargs in the signature, now here you may to pass those things explicitely as in `df.to_sql(con=create_engine(con, schema=schema, echo=False), **kwargs)`

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +299,65 @@ class DatabaseAsync(DatabaseView):
+    add_columns = ['database', 'schema', 'table_name']
+
+    def form_get(self, form):
+        form.sep.data = ','
```
I think the normal way of setting defaults for wtforms is to do it when creating the form objects. As in:
`BooleanField('name', default=True, ...)`

mistercrunch commented on superset/views/core.py:
```diff
@@ -298,6 +299,65 @@ class DatabaseAsync(DatabaseView):
+        form.decimal.data = '.'
+        form.error_bad_lines.data = False
+        form.if_exists.data = 'append'
+        all_datasources = db.session.query(
```
this one may have to stay here since you don't want to do it in module scope...

timifasubaa replied:
fixed

mistercrunch reviewed (changes requested):
We're starting to get somewhere, should we add some unit tests?

mistercrunch commented on superset/db_engine_specs.py:
```diff
@@ -686,6 +772,56 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
+
+        if not bucket_path:
+            logging.info("No upload bucket specified")
+            flash(
```
i18n

timifasubaa replied:
Fixed
